### PR TITLE
Build actions dag to evaluate missing defaults.

### PR DESCRIPTION
**Pull request by KochetovNicolai**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### CODE REVIEW

**Comment by alesapin** [comment]
```
E           helpers.client.QueryRuntimeException: Client failed! Return code: 32, stderr: [node1] 2021.02.09 16:40:58.607555 [ 94 ] <Fatal> BaseDaemon: ########################################
E           [node1] 2021.02.09 16:40:58.607644 [ 94 ] <Fatal> BaseDaemon: (version 21.3.1.5957, build id: 53E3F1E384A0F8D3FA8273CF3E53C65E1B8BEF59) (from thread 93) (query_id: 6ed85250-f335-4cb6-83c2-8d6dee32e6d4) Received signal Segmentation fault (11)
E           [node1] 2021.02.09 16:40:58.607680 [ 94 ] <Fatal> BaseDaemon: Address: NULL pointer. Access: read. Unknown si_code.
E           [node1] 2021.02.09 16:40:58.607723 [ 94 ] <Fatal> BaseDaemon: Stack trace: 0xe7ccfa8 0xf75d101 0xf75cae0 0xf727673 0xf6b61c2 0xf1c74b4 0xf1c672b 0xec23814 0xec2cfcb 0xec2d5ec 0xec2d669 0xf6de122 0xf71478b 0xf7145ac 0xf714c99 0xf7145ac 0xf714c99 0xf7145ac 0xf714cd9 0xf7145ac 0xf714cd9 0xf7181e0 0xf715c66 0xf7159bd 0xf72330e 0x8526caf 0x852a743 0x7f1229795609 0x7f12296bc293
E           [node1] 2021.02.09 16:40:58.607826 [ 94 ] <Fatal> BaseDaemon: 1. DB::DataTypeNullable::serializeTextQuoted(DB::IColumn const&, unsigned long, DB::WriteBuffer&, DB::FormatSettings const&) const @ 0xe7ccfa8 in /usr/bin/clickhouse
E           [node1] 2021.02.09 16:40:58.607900 [ 94 ] <Fatal> BaseDaemon: 2. DB::IRowOutputFormat::write(std::__1::vector<COW<DB::IColumn>::immutable_ptr<DB::IColumn>, std::__1::allocator<COW<DB::IColumn>::immutable_ptr<DB::IColumn> > > const&, unsigned long) @ 0xf75d101 in /usr/bin/clickhouse
E           [node1] 2021.02.09 16:40:58.607947 [ 94 ] <Fatal> BaseDaemon: 3. DB::IRowOutputFormat::consume(DB::Chunk) @ 0xf75cae0 in /usr/bin/clickhouse
E           [node1] 2021.02.09 16:40:58.607997 [ 94 ] <Fatal> BaseDaemon: 4. DB::IOutputFormat::write(DB::Block const&) @ 0xf727673 in /usr/bin/clickhouse
E           [node1] 2021.02.09 16:40:58.608068 [ 94 ] <Fatal> BaseDaemon: 5. DB::MaterializingBlockOutputStream::write(DB::Block const&) @ 0xf6b61c2 in /usr/bin/clickhouse
E           [node1] 2021.02.09 16:40:58.608123 [ 94 ] <Fatal> BaseDaemon: 6. DB::StorageMySQLBlockOutputStream::writeBlockData(DB::Block const&) @ 0xf1c74b4 in /usr/bin/clickhouse
E           [node1] 2021.02.09 16:40:58.608172 [ 94 ] <Fatal> BaseDaemon: 7. DB::StorageMySQLBlockOutputStream::write(DB::Block const&) @ 0xf1c672b in /usr/bin/clickhouse
E           [node1] 2021.02.09 16:40:58.608213 [ 94 ] <Fatal> BaseDaemon: 8. DB::PushingToViewsBlockOutputStream::write(DB::Block const&) @ 0xec23814 in /usr/bin/clickhouse
E           [node1] 2021.02.09 16:40:58.608277 [ 94 ] <Fatal> BaseDaemon: 9. DB::AddingDefaultBlockOutputStream::write(DB::Block const&) @ 0xec2cfcb in /usr/bin/clickhouse
E           [node1] 2021.02.09 16:40:58.608326 [ 94 ] <Fatal> BaseDaemon: 10. DB::SquashingBlockOutputStream::finalize() @ 0xec2d5ec in /usr/bin/clickhouse
E           [node1] 2021.02.09 16:40:58.608371 [ 94 ] <Fatal> BaseDaemon: 11. DB::SquashingBlockOutputStream::writeSuffix() @ 0xec2d669 in /usr/bin/clickhouse
E           [node1] 2021.02.09 16:40:58.608427 [ 94 ] <Fatal> BaseDaemon: 12. DB::ISink::prepare() @ 0xf6de122 in /usr/bin/clickhouse
E           [node1] 2021.02.09 16:40:58.608476 [ 94 ] <Fatal> BaseDaemon: 13. DB::PipelineExecutor::prepareProcessor(unsigned long, unsigned long, std::__1::queue<DB::ExecutingGraph::Node*, std::__1::deque<DB::ExecutingGraph::Node*, std::__1::allocator<DB::ExecutingGraph::Node*> > >&, std::__1::queue<DB::ExecutingGraph::Node*, std::__1::deque<DB::ExecutingGraph::Node*, std::__1::allocator<DB::ExecutingGraph::Node*> > >&, std::__1::unique_lock<std::__1::mutex>) @ 0xf71478b in /usr/bin/clickhouse
E           [node1] 2021.02.09 16:40:58.608531 [ 94 ] <Fatal> BaseDaemon: 14. DB::PipelineExecutor::tryAddProcessorToStackIfUpdated(DB::ExecutingGraph::Edge&, std::__1::queue<DB::ExecutingGraph::Node*, std::__1::deque<DB::ExecutingGraph::Node*, std::__1::allocator<DB::ExecutingGraph::Node*> > >&, std::__1::queue<DB::ExecutingGraph::Node*, std::__1::deque<DB::ExecutingGraph::Node*, std::__1::allocator<DB::ExecutingGraph::Node*> > >&, unsigned long) @ 0xf7145ac in /usr/bin/clickhouse
E           [node1] 2021.02.09 16:40:58.608582 [ 94 ] <Fatal> BaseDaemon: 15. DB::PipelineExecutor::prepareProcessor(unsigned long, unsigned long, std::__1::queue<DB::ExecutingGraph::Node*, std::__1::deque<DB::ExecutingGraph::Node*, std::__1::allocator<DB::ExecutingGraph::Node*> > >&, std::__1::queue<DB::ExecutingGraph::Node*, std::__1::deque<DB::ExecutingGraph::Node*, std::__1::allocator<DB::ExecutingGraph::Node*> > >&, std::__1::unique_lock<std::__1::mutex>) @ 0xf714c99 in /usr/bin/clickhouse
E           [node1] 2021.02.09 16:40:58.608640 [ 94 ] <Fatal> BaseDaemon: 16. DB::PipelineExecutor::tryAddProcessorToStackIfUpdated(DB::ExecutingGraph::Edge&, std::__1::queue<DB::ExecutingGraph::Node*, std::__1::deque<DB::ExecutingGraph::Node*, std::__1::allocator<DB::ExecutingGraph::Node*> > >&, std::__1::queue<DB::ExecutingGraph::Node*, std::__1::deque<DB::ExecutingGraph::Node*, std::__1::allocator<DB::ExecutingGraph::Node*> > >&, unsigned long) @ 0xf7145ac in /usr/bin/clickhouse
E           [node1] 2021.02.09 16:40:58.608690 [ 94 ] <Fatal> BaseDaemon: 17. DB::PipelineExecutor::prepareProcessor(unsigned long, unsigned long, std::__1::queue<DB::ExecutingGraph::Node*, std::__1::deque<DB::ExecutingGraph::Node*, std::__1::allocator<DB::ExecutingGraph::Node*> > >&, std::__1::queue<DB::ExecutingGraph::Node*, std::__1::deque<DB::ExecutingGraph::Node*, std::__1::allocator<DB::ExecutingGraph::Node*> > >&, std::__1::unique_lock<std::__1::mutex>) @ 0xf714c99 in /usr/bin/clickhouse
E           [node1] 2021.02.09 16:40:58.608759 [ 94 ] <Fatal> BaseDaemon: 18. DB::PipelineExecutor::tryAddProcessorToStackIfUpdated(DB::ExecutingGraph::Edge&, std::__1::queue<DB::ExecutingGraph::Node*, std::__1::deque<DB::ExecutingGraph::Node*, std::__1::allocator<DB::ExecutingGraph::Node*> > >&, std::__1::queue<DB::ExecutingGraph::Node*, std::__1::deque<DB::ExecutingGraph::Node*, std::__1::allocator<DB::ExecutingGraph::Node*> > >&, unsigned long) @ 0xf7145ac in /usr/bin/clickhouse
E           [node1] 2021.02.09 16:40:58.608826 [ 94 ] <Fatal> BaseDaemon: 19. DB::PipelineExecutor::prepareProcessor(unsigned long, unsigned long, std::__1::queue<DB::ExecutingGraph::Node*, std::__1::deque<DB::ExecutingGraph::Node*, std::__1::allocator<DB::ExecutingGraph::Node*> > >&, std::__1::queue<DB::ExecutingGraph::Node*, std::__1::deque<DB::ExecutingGraph::Node*, std::__1::allocator<DB::ExecutingGraph::Node*> > >&, std::__1::unique_lock<std::__1::mutex>) @ 0xf714cd9 in /usr/bin/clickhouse
E           [node1] 2021.02.09 16:40:58.608880 [ 94 ] <Fatal> BaseDaemon: 20. DB::PipelineExecutor::tryAddProcessorToStackIfUpdated(DB::ExecutingGraph::Edge&, std::__1::queue<DB::ExecutingGraph::Node*, std::__1::deque<DB::ExecutingGraph::Node*, std::__1::allocator<DB::ExecutingGraph::Node*> > >&, std::__1::queue<DB::ExecutingGraph::Node*, std::__1::deque<DB::ExecutingGraph::Node*, std::__1::allocator<DB::ExecutingGraph::Node*> > >&, unsigned long) @ 0xf7145ac in /usr/bin/clickhouse
E           [node1] 2021.02.09 16:40:58.608942 [ 94 ] <Fatal> BaseDaemon: 21. DB::PipelineExecutor::prepareProcessor(unsigned long, unsigned long, std::__1::queue<DB::ExecutingGraph::Node*, std::__1::deque<DB::ExecutingGraph::Node*, std::__1::allocator<DB::ExecutingGraph::Node*> > >&, std::__1::queue<DB::ExecutingGraph::Node*, std::__1::deque<DB::ExecutingGraph::Node*, std::__1::allocator<DB::ExecutingGraph::Node*> > >&, std::__1::unique_lock<std::__1::mutex>) @ 0xf714cd9 in /usr/bin/clickhouse
E           [node1] 2021.02.09 16:40:58.608995 [ 94 ] <Fatal> BaseDaemon: 22. DB::PipelineExecutor::executeStepImpl(unsigned long, unsigned long, std::__1::atomic<bool>*) @ 0xf7181e0 in /usr/bin/clickhouse
E           [node1] 2021.02.09 16:40:58.609036 [ 94 ] <Fatal> BaseDaemon: 23. DB::PipelineExecutor::executeImpl(unsigned long) @ 0xf715c66 in /usr/bin/clickhouse
E           [node1] 2021.02.09 16:40:58.609082 [ 94 ] <Fatal> BaseDaemon: 24. DB::PipelineExecutor::execute(unsigned long) @ 0xf7159bd in /usr/bin/clickhouse
E           [node1] 2021.02.09 16:40:58.609131 [ 94 ] <Fatal> BaseDaemon: 25. ? @ 0xf72330e in /usr/bin/clickhouse
E           [node1] 2021.02.09 16:40:58.609180 [ 94 ] <Fatal> BaseDaemon: 26. ThreadPoolImpl<std::__1::thread>::worker(std::__1::__list_iterator<std::__1::thread, void*>) @ 0x8526caf in /usr/bin/clickhouse
E           [node1] 2021.02.09 16:40:58.609226 [ 94 ] <Fatal> BaseDaemon: 27. ? @ 0x852a743 in /usr/bin/clickhouse
E           [node1] 2021.02.09 16:40:58.609266 [ 94 ] <Fatal> BaseDaemon: 28. start_thread @ 0x9609 in /usr/lib/x86_64-linux-gnu/libpthread-2.31.so
E           [node1] 2021.02.09 16:40:58.609322 [ 94 ] <Fatal> BaseDaemon: 29. __clone @ 0x122293 in /usr/lib/x86_64-linux-gnu/libc-2.31.so
E           [node1] 2021.02.09 16:40:58.759477 [ 94 ] <Fatal> BaseDaemon: Calculated checksum of the binary: 6B4E37488B04E3DC9FEEE56C79C4294E. There is no information about the reference checksum.
E           Error on processing query 'INSERT INTO test_insert_select(id, name, money, column_x) select number, concat('name_', toString(number)), 3, NULL from numbers(49) ':
E           Code: 32, e.displayText() = DB::Exception: Attempt to read after eof: while receiving packet from 172.18.0.4:9000, Stack trace (when copying this message, always include the lines below):
```

File: src/Functions/replicate.cpp [hunk 1]
```diff
@@ -11,60 +12,50 @@ namespace ErrorCodes
...
 }
 
-namespace
+DataTypePtr FunctionReplicate::getReturnTypeImpl(const DataTypes & arguments) const
 {
+    if (arguments.size() < 2)
+        throw Exception(ErrorCodes::TOO_FEW_ARGUMENTS_FOR_FUNCTION,
+                        "Function {} expect at least two arguments, got {}", getName(), arguments.size());
 
-/** Creates an array, multiplying the column (the first argument) by the number of elements in the array (the second argument).
```
Comment (alesapin) [inline src/Functions/replicate.cpp:19]: Lost comment

File: src/DataStreams/AddingDefaultBlockOutputStream.cpp [hunk 2]
```diff
@@ -1,13 +1,27 @@
...
 
+AddingDefaultBlockOutputStream::AddingDefaultBlockOutputStream(
+    const BlockOutputStreamPtr & output_,
+    const Block & header_,
+    const ColumnsDescription & columns_,
+    const Context & context_)
+    : output(output_), header(header_)
+{
+    auto dag = addMissingDefaults(header_, output->getHeader().getNamesAndTypesList(), columns_, context_);
+    actions = std::make_shared<ExpressionActions>(std::move(dag));
```
Comment (alesapin) [inline src/DataStreams/AddingDefaultBlockOutputStream.cpp:17]: `adding_defaults_actions`?

File: src/Interpreters/inplaceBlockConversions.h [hunk 3]
```diff
@@ -12,9 +13,12 @@ class Context;
...
 class ColumnsDescription;
 
-/// Adds missing defaults to block according to required_columns
-/// using columns description
-void evaluateMissingDefaults(Block & block,
+class ActionsDAG;
+using ActionsDAGPtr = std::shared_ptr<ActionsDAG>;
+
+/// Create actions which adds missing defaults to block according to required_columns using columns description.
+ActionsDAGPtr evaluateMissingDefaults(
```
Comment (alesapin) [inline src/Interpreters/inplaceBlockConversions.h:20]: Quite a confusing interface. "If no actions required then return nullptr". At least improve comment.

File: src/Interpreters/addMissingDefaults.cpp [hunk 4]
```diff
@@ -7,75 +7,94 @@
...
+
+            group.push_back(elem.name);
         }
     }
 
-    const size_t rows = block.rows();
-    Block res;
+    auto actions = std::make_shared<ActionsDAG>(header.getColumnsWithTypeAndName());
+
+    FunctionOverloadResolverPtr func_builder_replicate =
```
Comment (alesapin) [inline src/Interpreters/addMissingDefaults.cpp:47]: Move to private template function? Btw, looks scary.

Comment (alesapin) [reply src/Interpreters/addMissingDefaults.cpp:47]: Why we cannot get it from the factory?

File: src/Interpreters/addMissingDefaults.cpp [hunk 5]
```diff
@@ -7,75 +7,94 @@
...
+            std::make_shared<FunctionOverloadResolverAdaptor>(
+                    std::make_unique<DefaultOverloadResolver>(
+                            std::make_shared<FunctionReplicate>()));
+
+    FunctionOverloadResolverPtr func_builder_materialize =
```
Comment (alesapin) [inline src/Interpreters/addMissingDefaults.cpp:52]: Unused?

**Review by alesapin — COMMENTED** [review]
What is the main purpose of these changes? Seems like we want to get rid of manipulation with individual actions and do everything with DAG.

File: src/Interpreters/addMissingDefaults.cpp (hunk 4 above)
Comment (KochetovNicolai) [reply src/Interpreters/addMissingDefaults.cpp:47]: Well, here we can. Usually, I don't use factory cause it needs context

**Comment by KochetovNicolai** [comment]
> What is the main purpose of these changes? Seems like we want to get rid of manipulation with individual actions and do everything with DAG.

Exactly.
If something may be expressed in ActionsDAG, it most likely need to be expressed this way because:
* DAG may be build once, then executed on block stream. It should be cheaper then analyze each block.
* We have less processor types in pipeline and less step types in query plan
* DAG may be optimized

**Review by alesapin — APPROVED** [review]

**Pull request merged**